### ISSUE TITLE
Regression, and command-line agda typechecks less than interactive

### ISSUE THREAD

**Issue by GoogleCodeExporter** [agda/agda#1574]
With the development versions of agda and the standard library from yesterday 
(2015-06-16), using command-line agda does not typecheck the following modules, 
but in Emacs, it ``load''s without problems (at least after 
kill-and-restart-Agda --- in an active session, I saw the same problem 
occasionally, but could not reproduce it).

This typechecks without problems in Agda-2.4.2.3.

``` agda
===========================================
module StepLeq where

open import Data.Nat using ( ℕ ; zero ; suc ; _≤_ ; z≤n ; s≤s )

≤-step : {m n : ℕ} → m ≤ n → m ≤ suc n
≤-step z≤n = z≤n
≤-step (s≤s m≤n) = s≤s (≤-step m≤n)

step-≤ : {m n : ℕ} → suc m ≤ n → m ≤ n
step-≤ (s≤s p) = ≤-step p

{-
StepLeq.agda:10,1-26
I'm not sure if there should be a case for the constructor z≤n,
because I get stuck when trying to solve the following unification
problems (inferred index ≟ expected index):
  0 ≟ suc m
  n₁ ≟ n
when checking the definition of step-≤
-}
```

Original issue reported on code.google.com by `k...@cas.mcmaster.ca` on 18 Jun 2015 at 2:07

**Comment by UlfNorell** [comment]
We're not keeping the current sharing implementation, so it doesn't make sense to spend effort on this.

**Comment by asr** [comment]
> We're not keeping the current sharing implementation...

What does this mean? If it means that sharing won't be released with Agda 2.4.4, we should reopen the issue and remove the 2.4.4 milestone (which you did).

**Comment by asr** [comment]
Andreas answered my question in https://github.com/agda/agda/issues/1691#issuecomment-149049281:

> But anyway, the current sharing implementation with IORefs is not there to stay.

**Comment by andreasabel** [comment]
@UlfNorell s plan was to release 2.4.4 soon and he started to burn the bridges for maint-2.4.2 by carrying refactorings out on master only, so I guess a replacement of the current sharing implementation will not happen in 2.4.4.

**Comment by UlfNorell** [comment]
No that's not likely.

**Comment by WolframKahl** [comment]
(Andreas “brutal fix” for #1691 in commit 759a4dc500629270ec04c378a724a038217a439f  also “fixed” this. See also resulting typecking timings reported in #1691 .)

**Comment by asr** [comment]
> (Andreas “brutal fix” for #1691 in commit 759a4dc also “fixed” this.)

Are you sure? I couldn't reproduce the issue by reverting Andreas commit https://github.com/agda/agda/commit/759a4dc500629270ec04c378a724a038217a439f.

**Comment by WolframKahl** [comment]
On Mon, Oct 19, 2015 at 08:39:08AM -0700, Andrés Sicard-Ramírez wrote:

> > (Andreas “brutal fix” for #1691 in commit 759a4dc also “fixed” this.)
> 
> Are you sure?

Well, a few commits before that brutal fix,
the problems that made me submit #1574 were still there,
and with the brutal fix they are gone.
(Otherwise Main and NatSumProps would not have gone through
 from scratch with --sharing.)

Wolfram

**Comment by asr** [comment]
> @UlfNorell s plan was to release 2.4.4 soon and he started to burn the bridges for maint-2.4.2 by carrying refactorings out on master only.

From [GHC Weekly News - 2015/10/16](https://ghc.haskell.org/trac/ghc/blog/weekly20151016), GHC headquarters is planning to release GHC 7.8.5 and 7.10.3. Since we don't know who will be released first, Agda 2.4.4 or a new version of GHC, I suggest to continue carrying refactorings and fixing issues in both branches, i.e. maint-2.4.2 and master.

I just back-ported some stuff related to the test-suite, the LaTeX and HTML backends, and Travis into the maint branch.
